File: android/src/main/kotlin/com/imenuo/flutterleancloud/im.kt
package com.imenuo.flutterleancloud

import com.avos.avoscloud.AVQuery
import com.avos.avoscloud.im.v2.*
import com.avos.avoscloud.im.v2.callback.AVIMConversationCallback
import com.avos.avoscloud.im.v2.callback.AVIMConversationQueryCallback
import com.avos.avoscloud.im.v2.callback.AVIMMessagesQueryCallback
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import timber.log.Timber


internal fun AVIMConversation.toFlutterMap(): Map<String, Any?> {
    val obj = HashMap<String, Any?>()
    obj["conversationId"] = this.conversationId
    obj["members"] = this.members
    obj["lastMessage"] = this.lastMessage?.toFlutterMap()
    obj["lastMessageAt"] = this.lastMessageAt?.time
    obj["unreadMessagesCount"] = this.unreadMessagesCount
    return obj
}

internal fun AVIMMessage.toFlutterMap(): Map<String, Any?> {
    val obj = HashMap<String, Any?>()
    obj["content"] = this.content
    obj["conversationId"] = this.conversationId
    obj["from"] = this.from
    obj["messageId"] = this.messageId
    obj["timestamp"] = this.timestamp
    obj["deliveredAt"] = this.deliveredAt
    obj["updateAt"] = this.updateAt
    obj["status"] = this.messageStatus.statusCode
    return obj
}

internal fun AVIMException.toFlutterMap(): Map<String, Any?> {
    val obj = HashMap<String, Any?>()
    obj["code"] = code
    obj["appCode"] = appCode
    obj["message"] = message
    return obj
}

internal fun parseCachePolicy(cachePolicy: Int?): AVQuery.CachePolicy? {
    return when (cachePolicy) {
        0 -> AVQuery.CachePolicy.CACHE_ELSE_NETWORK
        1 -> AVQuery.CachePolicy.CACHE_ONLY
        2 -> AVQuery.CachePolicy.CACHE_THEN_NETWORK
        3 -> AVQuery.CachePolicy.IGNORE_CACHE
        4 -> AVQuery.CachePolicy.NETWORK_ELSE_CACHE
        5 -> AVQuery.CachePolicy.NETWORK_ONLY
        else -> null
    }
}

internal class ClientEventHandler(private val plugin: FlutterLeanCloudPlugin) : AVIMClientEventHandler() {
    override fun onConnectionResume(client: AVIMClient) {
        this.plugin.sendEvent(ChannelEvent.Success(hashMapOf(
                "event" to "avIMClient_clientEventHandler_onConnectionResumed",
                "data" to hashMapOf(
                        "clientId" to client.clientId
                )
        )))
    }

    override fun onConnectionPaused(client: AVIMClient) {
        this.plugin.sendEvent(ChannelEvent.Success(hashMapOf(
                "event" to "avIMClient_clientEventHandler_onConnectionPaused",
                "data" to hashMapOf(
                        "clientId" to client.clientId
                )
        )))
    }

    override fun onClientOffline(client: AVIMClient, reason: Int) {}
}

internal class ConversationEventHandler(private val plugin: FlutterLeanCloudPlugin) : AVIMConversationEventHandler() {
    override fun onInvited(client: AVIMClient?, conversation: AVIMConversation?, p2: String?) {}

    override fun onMemberJoined(p0: AVIMClient?, p1: AVIMConversation?, p2: MutableList<String>?, p3: String?) {}

    override fun onKicked(client: AVIMClient, conversation: AVIMConversation, kickedBy: String?) {
        Timber.d("onKicked(${client.clientId}, ${conversation.conversationId}, $kickedBy)")

        this.plugin.sendEvent(ChannelEvent.Success(hashMapOf(
                "event" to "avIMClient_conversationEventHandler_onKicked",
                "data" to hashMapOf(
                        "clientId" to client.clientId,
                        "conversation" to conversation.toFlutterMap(),
                        "kickedBy" to kickedBy
                )
        )))
    }

    override fun onMemberLeft(p0: AVIMClient?, p1: AVIMConversation?, p2: MutableList<String>?, p3: String?) {}

    override fun onUnreadMessagesCountUpdated(client: AVIMClient, conversation: AVIMConversation) {
        super.onUnreadMessagesCountUpdated(client, conversation)

        this.plugin.sendEvent(ChannelEvent.Success(hashMapOf(
                "event" to "avIMClient_conversationEventHandler_onUnreadMessagesCountUpdated",
                "data" to hashMapOf(
                        "clientId" to client.clientId,
                        "conversation" to conversation.toFlutterMap()
                )
        )))
    }
}

private class MessageHandler(private val plugin: FlutterLeanCloudPlugin) : AVIMMessageHandler() {
    override fun onMessage(message: AVIMMessage, conversation: AVIMConversation, client: AVIMClient) {
        super.onMessage(message, conversation, client)

        this.plugin.sendEvent(ChannelEvent.Success(hashMapOf(
                "event" to "avIMClient_messageHandler_onMessage",
                "data" to hashMapOf(
                        "clientId" to client.clientId,
                        "message" to message.toFlutterMap(),
                        "conversation" to conversation.toFlutterMap()
                )
        )))
    }
}

internal class AVIMClientMethodCallHandler(private val plugin: FlutterLeanCloudPlugin) : SubMethodCallHandler {
    private var messageHandler: MessageHandler? = null

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result): Boolean {
        if (!call.method.startsWith("avIMClient_"))
            return false

        val method = call.method.removePrefix("avIMClient_")
        when (method) {
            "registerMessageHandler" -> registerMessageHandler(result)
            "unregisterMessageHandler" -> unregisterMessageHandler(result)
            "getInstance" -> getInstance(call, result)
            "queryConversations" -> queryConversations(call, result)
            else -> throw NotImplementedError("unimplemented handler for: ${call.method}")
        }

        return true
    }

    private fun queryConversations(call: MethodCall, result: MethodChannel.Result) {
        val clientId: String = call.argument("clientId")!!
        val client = AVIMClient.getInstance(clientId)
        val ids: List<String> = call.argument("ids")!!
        // TODO 2018-08-19: isCompact not implemented in AVIMConversationsQuery, but documented in website
        // val isCompact: Boolean = call.argument("isCompact")
        val refreshLastMessage: Boolean = call.argument("refreshLastMessage")!!
        val cachePolicy: AVQuery.CachePolicy? = parseCachePolicy(call.argument("cachePolicy"))
        val limit: Int = call.argument("limit")!!

        val query = client.conversationsQuery
        query.limit(limit)
        query.whereContainsIn("objectId", ids)
        query.isWithLastMessagesRefreshed = refreshLastMessage
        if (cachePolicy != null) query.setQueryPolicy(cachePolicy)
        query.findInBackground(object : AVIMConversationQueryCallback() {
            override fun done(conversations: MutableList<AVIMConversation>?, exception: AVIMException?) {
                if (exception != null) {
                    Timber.e(exception, "cannot query conversations")
                    result.error("ERROR", "cannot query conversations", exception.toString())
                    return
                }
                if (conversations == null) {
                    result.error("ERROR", "assertion failed: conversation should not be null", null)
                    return
                }
                result.success(conversations.map { it.toFlutterMap() }.toMutableList())
            }
        })
    }

    private fun registerMessageHandler(result: MethodChannel.Result) {
        if (messageHandler != null) {
            result.success(null)
            return
        }
        messageHandler = MessageHandler(this.plugin)
        AVIMMessageManager.registerMessageHandler(AVIMMessage::class.java, messageHandler)
        result.success(null)
    }

    private fun unregisterMessageHandler(result: MethodChannel.Result) {
        if (messageHandler == null) {
            result.success(null)
            return
        }
        AVIMMessageManager.unregisterMessageHandler(AVIMMessage::class.java, messageHandler)
        messageHandler = null
        result.success(null)
    }

    private fun getInstance(call: MethodCall, result: MethodChannel.Result) {
        AVIMClient.getInstance(call.arguments as String)
        result.success(null)
    }
}

internal class AVIMConversationMethodCallHandler(private val plugin: FlutterLeanCloudPlugin) : SubMethodCallHandler {
    private fun internalGetConversation(clientId: String, conversationId: String): AVIMConversation {
        return AVIMClient.getInstance(clientId).getConversation(conversationId)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result): Boolean {
        if (!call.method.startsWith("avIMConversation_"))
            return false

        val method = call.method.removePrefix("avIMConversation_")
        when (method) {
            "queryMessages" -> queryMessages(call, result)
            "sendMessage" -> sendMessage(call, result)
            "read" -> read(call, result)
            else -> throw NotImplementedError("unimplemented handler for ${call.method}")
        }

        return true
    }

    private fun sendMessage(call: MethodCall, result: MethodChannel.Result) {
        val clientId: String = call.argument("clientId")!!
        val conversationId: String = call.argument("conversationId")!!
        val content: String = call.argument("content")!!

        val conversation = internalGetConversation(clientId, conversationId)
        val message = AVIMMessage()
        message.content = content
        conversation.sendMessage(message, object : AVIMConversationCallback() {
            override fun done(exception: AVIMException?) {
                if (exception != null) {
                    val data = HashMap<String, Any?>()
                    data["message"] = message.toFlutterMap()
                    data["exception"] = exception.toFlutterMap()
                    result.error("ERROR", "send message failed", data)
                    return
                }
                result.success(message.toFlutterMap())
            }
        })
    }

    private fun queryMessages(call: MethodCall, result: MethodChannel.Result) {
        val clientId: String = call.argument("clientId")!!
        val conversationId: String = call.argument("conversationId")!!
        val msgId: String? = call.argument("msgId")
        val timestamp: Long? = call.argument("timestamp")
        val limit: Int = call.argument("limit")!!

        val conversation = internalGetConversation(clientId, conversationId)
        if (msgId == null || timestamp == null) {
            conversation.queryMessages(limit, MessagesQueryCallback(result))
        } else {
            conversation.queryMessages(msgId, timestamp, limit, MessagesQueryCallback(result))
        }
    }

    private fun read(call: MethodCall, result: MethodChannel.Result) {
        val clientId: String = call.argument("clientId")!!
        val conversationId: String = call.argument("conversationId")!!

        val conversation = internalGetConversation(clientId, conversationId)
        conversation.read()
        result.success(null)
    }

    class MessagesQueryCallback(private val result: MethodChannel.Result) : AVIMMessagesQueryCallback() {
        override fun done(msgs: MutableList<AVIMMessage>?, exception: AVIMException?) {
            if (exception != null) {
                result.error("ERROR", "cannot query messages", exception.toString())
                return
            }
            if (msgs == null) {
                result.error("ERROR", "assertion failed: msgs should not be null", null)
                return
            }
            result.success(msgs.map { it.toFlutterMap() }.toMutableList())
        }
    }
}

File: ios/Classes/IM.swift
//
// Created by Kainan Zhu on 2018/8/30.
//

import Foundation
import AVOSCloud
import AVOSCloudIM

extension AVIMConversation {
    func toFlutterDictionary() -> [String: Any?] {
        return [
            "conversationId": self.conversationId,
            "members": self.members,
            "lastMessage": self.lastMessage?.toFlutterDictionary(),
            "lastMessageAt": self.lastMessageAt?.millisecondsSince1970,
            "unreadMessagesCount": self.unreadMessagesCount,
        ]
    }
}

extension AVIMMessageStatus {
    var statusCode: Int? {
        get {
            switch self {
            case .none:
                return 0
            case .sending:
                return 1
            case .sent:
                return 2
            case .delivered:
                return 3
            case .failed:
                return 4
            default:
                return nil
            }
        }
    }
}

extension AVIMMessage {
    func toFlutterDictionary() -> [String: Any?] {
        return [
            "content": self.content,
            "conversationId": self.conversationId,
            "from": self.clientId, // TODO add this to document
            "messageId": self.messageId,
            "timestamp": self.sendTimestamp,
            "deliveredAt": self.deliveredTimestamp,
            "updateAt": self.updatedAt?.millisecondsSince1970,
            "status": self.status.statusCode,
        ]
    }
}

extension AVIMCachePolicy {
    static func fromFlutterInt(_ value: Int?) -> AVIMCachePolicy? {
        switch value {
        case 0:
            return AVIMCachePolicy.cacheElseNetwork
        case 1:
            return AVIMCachePolicy.cacheOnly
        case 2:
            return AVIMCachePolicy.cacheThenNetwork
        case 3:
            return AVIMCachePolicy.ignoreCache
        case 4:
            return AVIMCachePolicy.networkElseCache
        case 5:
            return AVIMCachePolicy.networkOnly
        default:
            return nil
        }
    }
}

class ClientDelegate: NSObject, AVIMClientDelegate {
    private let plugin: SwiftFlutterLeanCloudPlugin

    init(with plugin: SwiftFlutterLeanCloudPlugin) {
        self.plugin = plugin
    }

    func imClientPaused(_ imClient: AVIMClient) {
        let data: [String: Any?] = [
            "clientId": imClient.clientId
        ]

        self.plugin.channel.invokeMethod("avIMClient_clientEventHandler_onConnectionPaused", arguments: data)
    }

    func imClientResuming(_ imClient: AVIMClient) {
        imClientResumed(imClient)
    }

    func imClientResumed(_ imClient: AVIMClient) {
        let data: [String: Any?] = [
            "clientId": imClient.clientId,
        ]

        self.plugin.channel.invokeMethod("avIMClient_clientEventHandler_onConnectionResumed", arguments: data)
    }

    func imClientClosed(_ imClient: AVIMClient, error: Error?) {
    }

    func conversation(_ conversation: AVIMConversation, didReceiveCommonMessage message: AVIMMessage) {
        let data: [String: Any?] = [
            "clientId": conversation.clientId,
            "message": message.toFlutterDictionary(),
            "conversation": conversation.toFlutterDictionary(),
        ]

        self.plugin.channel.invokeMethod("avIMClient_messageHandler_onMessage", arguments: data)
    }

    func conversation(_ conversation: AVIMConversation, didUpdateForKey key: AVIMConversationUpdatedKey) {
        if key == AVIMConversationUpdatedKey.unreadMessagesCount {
            let data: [String: Any?] = [
                "clientId": conversation.clientId,
                "conversation": conversation.toFlutterDictionary(),
            ]

            self.plugin.channel.invokeMethod("avIMClient_conversationEventHandler_onUnreadMessagesCountUpdated",
                    arguments: data)
        }
    }
}

class AVIMClientMethodCallHandler: SubMethodCallHandler {
    private static var clientCache: [String: AVIMClient] = [:]
    
    private let plugin: SwiftFlutterLeanCloudPlugin
    private let clientDelegate: AVIMClientDelegate
  
    static func getClient(clientId: String) -> AVIMClient {
        var client = clientCache[clientId]
        if client == nil {
            client = AVIMClient.init(clientId: clientId)
            clientCache[clientId] = client
        }
        return client!
    }
    
    init(with plugin: SwiftFlutterLeanCloudPlugin) {
        self.plugin = plugin
        self.clientDelegate = ClientDelegate(with: plugin)
    }

    func onMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) -> Bool {
        if !call.method.starts(with: "avIMClient_") {
            return false
        }

        let method = call.method["avIMClient_".endIndex...]
        switch method {
        case "registerMessageHandler":
            result(nil)
        case "unregisterMessageHandler":
            result(nil)
        case "getInstance":
            getInstance(call, result)
        case "queryConversations":
            queryConversations(call, result)
        default:
            result(FlutterMethodNotImplemented)
        }

        return true
    }

    private func getInstance(_ call: FlutterMethodCall, _ result: (Any?) -> Void) {
        if call.arguments == nil {
            return result(FlutterError(code: "INVALID_ARGS", message: "clientId cannot be nil", details: nil))
        }
        let clientId = call.arguments as! String
        let client = AVIMClientMethodCallHandler.getClient(clientId: clientId)
        client.delegate = self.clientDelegate
        result(nil)
    }

    private func queryConversations(_ call: FlutterMethodCall, _ result: @escaping (Any?) -> Void) {
        let args = call.arguments as! [String: Any?]
        let clientId = args["clientId"] as! String
        let client = AVIMClientMethodCallHandler.getClient(clientId: clientId)
        let ids = args["ids"] as! [String]
        let refreshLastMessage = args["refreshLastMessage"] as! Bool
        let cachePolicy = AVIMCachePolicy.fromFlutterInt(args["cachePolicy"] as! Int?)
        let limit = args["limit"] as! Int

        let query = client.conversationQuery()
        query.limit = limit
        query.whereKey("objectId", containedIn: ids)
        if refreshLastMessage {
            query.option = AVIMConversationQueryOption.withMessage
        }
        if let cachePolicy = cachePolicy {
            query.cachePolicy = cachePolicy
        }
        query.findConversations(callback: { (conversations, error) in
            if let error = error {
                return result(FlutterError(code: "ERROR", message: "cannot query conversations",
                        details: error.localizedDescription))
            }
            if conversations == nil {
                return result(FlutterError(code: "ERROR", message: "assertion failed: conversation should not be null",
                        details: nil))
            }
            result(conversations?.map({ $0.toFlutterDictionary() }))
        })
    }
}

class AVIMConversationMethodCallHandler: SubMethodCallHandler {
    private func internalGetConversation(in clientId: String, forId conversationId: String,
                                         callback: @escaping (AVIMConversation?, Error?) -> Void) -> Void {
        let client = AVIMClientMethodCallHandler.getClient(clientId: clientId)
        let conversation = client.conversation(forId: conversationId)
        if let conversation = conversation {
            debugPrint("got AVIMConversation from cache")
            return callback(conversation, nil)
        }
        let query = client.conversationQuery()
        query.getConversationById(conversationId, callback: { (conversation, error) in
            debugPrint("got AVIMConversation from network")
            callback(conversation, error)
        })
    }

    func onMethodCall(_ call: FlutterMethodCall, result: @escaping FlutterResult) -> Bool {
        if !call.method.starts(with: "avIMConversation_") {
            return false
        }

        let method = call.method["avIMConversation_".endIndex...]
        switch method {
        case "queryMessages":
            queryMessages(call, result)
        case "sendMessage":
            sendMessage(call, result)
        case "read":
            read(call, result)
        default:
            result(FlutterMethodNotImplemented)
        }

        return true
    }

    private func read(_ call: FlutterMethodCall, _ result: @escaping (Any?) -> Void) {
        let args = call.arguments as! [String: Any?]
        let clientId = args["clientId"] as! String
        let conversationId = args["conversationId"] as! String

        internalGetConversation(in: clientId, forId: conversationId, callback: { (conversation, error) in
            if let error = error {
                return result(FlutterError(code: "ERROR", message: "cannot get conversation", details: error.localizedDescription))
            }
            if let conversation = conversation {
                conversation.readInBackground()
                return result(nil)
            } else {
                return result(FlutterError(code: "ERROR", message: "cannot get conversation", details: nil))
            }
        })
    }

    private func sendMessage(_ call: FlutterMethodCall, _ result: @escaping (Any?) -> Void) {
        let args = call.arguments as! [String: Any?]
        let clientId = args["clientId"] as! String
        let conversationId = args["conversationId"] as! String
        let content = args["content"] as! String

        internalGetConversation(in: clientId, forId: conversationId, callback: { (conversation, error) in
            if let error = error {
                return result(FlutterError(code: "ERROR", message: "cannot get conversation", details: error.localizedDescription))
            }
            if let conversation = conversation {
                let message = AVIMMessage(content: content)
                conversation.send(message, callback: { (succeeded, error) in
                    if let error = error {
                        let error = error as NSError
                        let exception: [String: Any?] = [
                            "code": error.code,
                            "appCode": error.code,
                            "message": error.userInfo[NSLocalizedFailureReasonErrorKey],
                        ]
                        let data: [String: Any?] = [
                            "message": message.toFlutterDictionary(),
                            "exception": exception,
                        ]
                        return result(FlutterError(code: "ERROR", message: "send message failed", details: data))
                    }
                    return result(message.toFlutterDictionary())
                })
            } else {
                return result(FlutterError(code: "ERROR", message: "cannot get conversation", details: nil))
            }
        })
    }

    private func queryMessages(_ call: FlutterMethodCall, _ result: @escaping (Any?) -> Void) {
        let args = call.arguments as! [String: Any?]
        let clientId = args["clientId"] as! String
        let conversationId = args["conversationId"] as! String
        let msgId = args["msgId"] as! String?
        let timestamp = args["timestamp"] as! Int64?
        let limit = args["limit"] as! Int

        func callback(_ messages: [AVIMMessage]?, _ error: Error?) {
            if let error = error {
                return result(FlutterError(code: "ERROR", message: "cannot query messages",
                        details: error.localizedDescription))
            }
            if let messages = messages {
                return result(messages.map({ $0.toFlutterDictionary() }))
            } else {
                return result(FlutterError(code: "ERROR", message: "assertion failed: msgs should not be nil",
                        details: nil))
            }
        }

        internalGetConversation(in: clientId, forId: conversationId, callback: { (conversation, error) in
            if let error = error {
                return result(FlutterError(code: "ERROR", message: "cannot get conversation",
                        details: error.localizedDescription))
            }
            if let conversation = conversation {
                if let msgId = msgId, let timestamp = timestamp {
                    conversation.queryMessages(beforeId: msgId, timestamp: timestamp, limit: UInt(limit),
                            callback: callback)
                } else {
                    conversation.queryMessages(withLimit: UInt(limit), callback: callback)
                }
            } else {
                return result(FlutterError(code: "ERROR", message: "cannot get conversation", details: nil))
            }
        })
    }
}
